File: Javascript contract deployment env/scripts/interaction2.js
require('dotenv').config();
const { ethers } = require('hardhat');
const fs = require('fs');
const path = require('path');

async function main() {
    // Setup deployer using the private key from .env
    const privateKey = process.env.PRIVATE_KEY;
    if (!privateKey) {
        throw new Error('Please set your PRIVATE_KEY in a .env file');
    }

    const provider = new ethers.providers.JsonRpcProvider();
    const deployer = new ethers.Wallet(privateKey, provider);

    console.log('Interacting with contracts using the account:', deployer.address);

    // Load deployed addresses
    const addresses = JSON.parse(fs.readFileSync('deployedAddresses.json'));

    const ActivityMint = await ethers.getContractFactory('Activity_Mint');
    const aCoins = ActivityMint.attach(addresses.aCoins);

    const wethArtifactPath = path.join(__dirname, '../contracts/WETH9.json');
    const routerArtifactPath = path.join(__dirname, '../contracts/UniswapV2Router02.json');
    const pairArtifactPath = path.join(__dirname, '../contracts/UniswapV2Pair.json');

    const wethArtifact = JSON.parse(fs.readFileSync(wethArtifactPath));
    const routerArtifact = JSON.parse(fs.readFileSync(routerArtifactPath));
    const pairArtifact = JSON.parse(fs.readFileSync(pairArtifactPath));

    const weth = new ethers.Contract(addresses.weth, wethArtifact.abi, deployer);
    const router = new ethers.Contract(addresses.router, routerArtifact.abi, deployer);
    const pair = new ethers.Contract(addresses.pair, pairArtifact.abi, deployer);

    // Mint WETH
    const mintWethTx = await weth.deposit({ value: ethers.utils.parseUnits('2', 18) });
    await mintWethTx.wait();
    console.log('Minted 2 WETH');

    // Display deployer's balance
    const aCoinsBalance = await aCoins.balanceOf(deployer.address);
    const wethBalance = await weth.balanceOf(deployer.address);
    console.log('Deployer ACoins balance:', ethers.utils.formatUnits(aCoinsBalance, 18));
    console.log('Deployer WETH balance:', ethers.utils.formatUnits(wethBalance, 18));

    // Add liquidity
    const aCoinsAmount = ethers.utils.parseUnits('7500', 18); // 7500 ACoins
    const wethAmount = ethers.utils.parseUnits('2', 18); // 2 WETH
    const deadline = Math.floor(Date.now() / 1000) + (10 * 60); // 10 minutes from the current Unix time

    // Approve tokens for liquidity addition
    await aCoins.approve(router.address, aCoinsAmount).then(tx => tx.wait());
    await weth.approve(router.address, wethAmount).then(tx => tx.wait());

    const addLiquidityTx = await router.addLiquidity(
        aCoins.address,
        weth.address,
        aCoinsAmount,
        wethAmount,
        0,
        0,
        deployer.address,
        deadline,
        { gasLimit: ethers.utils.hexlify(1000000) }
    );
    await addLiquidityTx.wait();
    console.log('Liquidity added');

    // Fetch and display reserves before the swap
    let reserves = await pair.getReserves();
    console.log('Reserves before swap:', reserves);
    console.log('Pair ACoins balance before swap:', ethers.utils.formatUnits(reserves._reserve1, 18));
    console.log('Pair WETH balance before swap:', ethers.utils.formatUnits(reserves._reserve0, 18));

    // Perform a swap from ACoins to WETH
    const amountInACoins = ethers.utils.parseUnits('500', 18); // 500 ACoins
    const amountOutMinWeth = 0; // Accept any amount of WETH

    await aCoins.approve(router.address, amountInACoins).then(tx => tx.wait());
    const swapPathACoinsToWeth = [aCoins.address, weth.address];
    const aCoinsToWethSwapTx = await router.swapExactTokensForTokens(
        amountInACoins,
        amountOutMinWeth,
        swapPathACoinsToWeth,
        deployer.address,
        deadline,
        { gasLimit: ethers.utils.hexlify(1000000) }
    );
    await aCoinsToWethSwapTx.wait();
    console.log('ACoins to WETH swap executed');

    // Fetch and display reserves after the first swap
    reserves = await pair.getReserves();
    console.log('Reserves after ACoins to WETH swap:', reserves);
    console.log('Pair ACoins balance after ACoins to WETH swap:', ethers.utils.formatUnits(reserves._reserve1, 18));
    console.log('Pair WETH balance after ACoins to WETH swap:', ethers.utils.formatUnits(reserves._reserve0, 18));

    // Perform a swap from WETH to ACoins
    const amountInWeth = ethers.utils.parseUnits('0.2', 18); // 0.2 WETH
    const amountOutMinACoins = 0; // Accept any amount of ACoins

    await weth.approve(router.address, amountInWeth).then(tx => tx.wait());
    const swapPathWethToACoins = [weth.address, aCoins.address];
    const wethToACoinsSwapTx = await router.swapExactTokensForTokens(
        amountInWeth,
        amountOutMinACoins,
        swapPathWethToACoins,
        deployer.address,
        deadline,
        { gasLimit: ethers.utils.hexlify(1000000) }
    );
    await wethToACoinsSwapTx.wait();
    console.log('WETH to ACoins swap executed');

    // Fetch and display reserves after the second swap
    reserves = await pair.getReserves();
    console.log('Reserves after WETH to ACoins swap:', reserves);
    console.log('Pair ACoins balance after WETH to ACoins swap:', ethers.utils.formatUnits(reserves._reserve1, 18));
    console.log('Pair WETH balance after WETH to ACoins swap:', ethers.utils.formatUnits(reserves._reserve0, 18));

    // Display deployer's new balance
    const newACoinsBalance = await aCoins.balanceOf(deployer.address);
    const newWethBalance = await weth.balanceOf(deployer.address);
    console.log('New ACoins balance:', ethers.utils.formatUnits(newACoinsBalance, 18));
    console.log('New WETH balance:', ethers.utils.formatUnits(newWethBalance, 18));
}

main()
    .then(() => process.exit(0))
    .catch((error) => {
        console.error(error);
        process.exit(1);
    });

File: EVM Blockchain/scripts/interaction2.py
from brownie import accounts, network, web3, Activity_Mint, interface
import json
import os
from dotenv import load_dotenv
from pathlib import Path

def main():
    # Load environment variables
    load_dotenv()
    gas_price = 5 * 10**9  # 2 gwei in wei
    gas_limit = 3000000  # Adjust as necessary

    # Set up the deployer account
    deployer = accounts.load('test2')
    print('Interacting with contracts using the account:', deployer.address)

    # Load the Activity_Mint address from mintaddy.txt
    with open('Activity-Mint.txt', 'r') as f:
        mint_address = f.read().strip()

    # Load deployed addresses
    with open('deployedAddresses.json') as f:
        addresses = json.load(f)

    # Load the pair address from coinpair.txt
    with open('coinpair.txt', 'r') as f:
        pair_address = f.read().strip()

    # Set up the contract objects
    ACoin = Activity_Mint.at(mint_address)
    router = interface.IUniswapV2Router02(addresses['router'])
    pair = interface.IUniswapV2Pair(pair_address)
    weth = interface.IWETH(addresses['weth'])

    # Step 1: Mint WETH
    mint_weth_tx = weth.deposit({'from': deployer, 'value': web3.to_wei(2, 'ether'), 'gas_price': gas_price, 'gas_limit': gas_limit})
    mint_weth_tx.wait(1)
    print('Minted 2 WETH')

    # Display deployer's balance
    ACoin_balance = ACoin.balanceOf(deployer.address)
    weth_balance = weth.balanceOf(deployer.address)
    print('Deployer ACoins balance:', web3.from_wei(ACoin_balance, 'ether'))
    print('Deployer WETH balance:', web3.from_wei(weth_balance, 'ether'))

    # Step 2: Add Liquidity
    aCoins_amount = web3.to_wei(7500, 'ether')  # 7500 ACoins
    weth_amount = web3.to_wei(2, 'ether')  # 2 WETH
    deadline = web3.eth.get_block('latest')['timestamp'] + (10 * 60)  # 10 minutes from the current Unix time

    # Approve tokens for liquidity addition
    ACoin.approve(router.address, aCoins_amount, {'from': deployer, 'gas_price': gas_price, 'gas_limit': gas_limit}).wait(1)
    weth.approve(router.address, weth_amount, {'from': deployer, 'gas_price': gas_price, 'gas_limit': gas_limit}).wait(1)

    add_liquidity_tx = router.addLiquidity(
        ACoin.address,
        weth.address,
        aCoins_amount,
        weth_amount,
        0,
        0,
        deployer.address,
        deadline,
        {'from': deployer, 'gas_price': gas_price, 'gas_limit': gas_limit}
    )
    add_liquidity_tx.wait(1)
    print('Liquidity added')

    # Step 3: Perform Swaps

    # Fetch and display reserves before the swap
    reserves = pair.getReserves()
    print('Reserves before swap:', reserves)
    print('Pair ACoins balance before swap:', web3.from_wei(reserves[0], 'ether'))
    print('Pair WETH balance before swap:', web3.from_wei(reserves[1], 'ether'))

    # Perform a swap from ACoins to WETH
    amount_in_aCoins = web3.to_wei(500, 'ether')  # 500 ACoins
    amount_out_min_weth = 0  # Accept any amount of WETH

    ACoin.approve(router.address, amount_in_aCoins, {'from': deployer, 'gas_price': gas_price, 'gas_limit': gas_limit}).wait(1)
    swap_path_aCoins_to_weth = [ACoin.address, weth.address]
    aCoins_to_weth_swap_tx = router.swapExactTokensForTokens(
        amount_in_aCoins,
        amount_out_min_weth,
        swap_path_aCoins_to_weth,
        deployer.address,
        (web3.eth.get_block('latest')['timestamp'] + (10 * 60)),
        {'from': deployer, 'gas_price': gas_price, 'gas_limit': gas_limit}
    )
    aCoins_to_weth_swap_tx.wait(1)
    print('ACoins to WETH swap executed')

    # Fetch and display reserves after the first swap
    reserves = pair.getReserves()
    print('Reserves after ACoins to WETH swap:', reserves)
    print('Pair ACoins balance after ACoins to WETH swap:', web3.from_wei(reserves[0], 'ether'))
    print('Pair WETH balance after ACoins to WETH swap:', web3.from_wei(reserves[1], 'ether'))

    # Perform a swap from WETH to ACoins
    amount_in_weth = web3.to_wei(0.1, 'ether')  # 0.1 WETH
    amount_out_min_aCoins = 0  # Accept any amount of ACoins

    weth.approve(router.address, amount_in_weth, {'from': deployer, 'gas_price': gas_price, 'gas_limit': gas_limit}).wait(1)
    swap_path_weth_to_aCoins = [weth.address, ACoin.address]
    weth_to_aCoins_swap_tx = router.swapExactTokensForTokens(
        amount_in_weth,
        amount_out_min_aCoins,
        swap_path_weth_to_aCoins,
        deployer.address,
        (web3.eth.get_block('latest')['timestamp'] + (10 * 60)),
        {'from': deployer, 'gas_price': gas_price, 'gas_limit': gas_limit}
    )
    weth_to_aCoins_swap_tx.wait(1)
    print('WETH to ACoins swap executed')

    # Fetch and display reserves after the second swap
    reserves = pair.getReserves()
    print('Reserves after WETH to ACoins swap:', reserves)
    print('Pair ACoins balance after WETH to ACoins swap:', web3.from_wei(reserves[0], 'ether'))
    print('Pair WETH balance after WETH to ACoins swap:', web3.from_wei(reserves[1], 'ether'))

    # Display deployer's new balance
    new_aCoins_balance = ACoin.balanceOf(deployer.address)
    new_weth_balance = weth.balanceOf(deployer.address)
    print('New ACoins balance:', web3.from_wei(new_aCoins_balance, 'ether'))
    print('New WETH balance:', web3.from_wei(new_weth_balance, 'ether'))

def load_dotenv():
    from dotenv import load_dotenv
    load_dotenv()

if __name__ == "__main__":
    main()
